common/types/priorities.ts

```diff
@@ -0,0 +1,94 @@
+export type PrioritiesT = readonly {readonly stage: string}[]
+export type Priority<T extends PrioritiesT> = number & {
+	__priority_type_do_not_use_or_program_will_crash: T
+}
+export type PriorityDict<T extends PrioritiesT> = {
+	readonly [stage in T[number]['stage']]: Priority<T>
+}
+export type PrioritySrc<T> = T extends PriorityDict<infer Priorities>
+	? Priorities
+	: never
+
+function createPriorityDictionary<T extends PrioritiesT>(
+	priorities: Exclude<T, {stage: string}[]>,
+): PriorityDict<T> {
+	return priorities.reduce(
+		(r, {stage}, i) => {
+			if (stage in r)
+				throw new Error(
+					'Priority dictionary should not include duplicate stages',
+				)
+			r[stage] = i
+			return r
+		},
+		{} as Record<string, number>,
+	) as any
+}
+
+/**
+ * @member PUSH_IGNORE_CARDS: rare xB, Ignore Attach effect
+ * @member HERMIT_SET_TARGET: Hermits updating the intended target of attacks
+ * @member HERMIT_CHANGE_TARGET: Betrayed effect, Sheep Stare effect
+ * @member LIGHTNING_ROD_REDIRECT
+ * @member TARGET_BLOCK_REDIRECT
+ * @member HERMIT_MODIFY_DAMAGE: Hermits changing the damage of an attack or adding sub-attacks
+ * @member EFFECT_MODIFY_DAMAGE: Invisibility effects
+ * @member HERMIT_APPLY_ATTACK: Hermits applying their effect to the board
+ * @member APPLY_SINGLE_USE_ATTACK: Single use attack cards: Bow, Swords, etc.
+ */
+export const beforeAttack = createPriorityDictionary([
+	{stage: 'PUSH_IGNORE_CARDS'},
+	{stage: 'HERMIT_SET_TARGET'},
+	{stage: 'HERMIT_CHANGE_TARGET'},
+	{stage: 'LIGHTNING_ROD_REDIRECT'},
+	{stage: 'TARGET_BLOCK_REDIRECT'},
+	{stage: 'HERMIT_MODIFY_DAMAGE'},
+	{stage: 'EFFECT_MODIFY_DAMAGE'},
+	{stage: 'HERMIT_APPLY_ATTACK'},
+	{stage: 'APPLY_SINGLE_USE_ATTACK'},
+] as const)
```

Comment:
This style of dictionary doesn't work with the LSP goto definition and goto references features.
It will be better to do this:

```
const priorities = createPriorityDictionary({
	'HELLO_WORLD': null,
	'SOMETHING_ELSE': null,
})

```

I did this for client and server references if you want to look at an example.